scene/3d/navigation_mesh_instance.cpp

```diff
@@ -0,0 +1,258 @@
+	}
+
+	emit_signal("navigation_mesh_changed");
+
+	update_gizmo();
+	update_configuration_warning();
+}
+
+Ref<NavigationMesh> NavigationMeshInstance::get_navigation_mesh() const {
+
+	return navmesh;
+}
+
+struct BakeThreadsArgs {
+	NavigationMeshInstance *nav_mesh_instance;
+};
+
+void _bake_navigation_mesh(void *p_user_data) {
+	BakeThreadsArgs *args = static_cast<BakeThreadsArgs *>(p_user_data);
+
+	if (args->nav_mesh_instance->get_navigation_mesh().is_valid()) {
+		Ref<NavigationMesh> nav_mesh = args->nav_mesh_instance->get_navigation_mesh()->duplicate();
+
+		NavigationServer::get_singleton()->region_bake_navmesh(nav_mesh, args->nav_mesh_instance);
+		args->nav_mesh_instance->call_deferred("_bake_done", nav_mesh);
+		memdelete(args);
+	} else {
+
+		ERR_PRINT("Can't bake the navigation mesh if the `NavigationMesh` resource doesn't exist");
+		args->nav_mesh_instance->call_deferred("_bake_done", Ref<NavigationMesh>());
+		memdelete(args);
+	}
+}
+
+void NavigationMeshInstance::bake_navigation_mesh() {
+	ERR_FAIL_COND(bake_thread != NULL);
+
+	BakeThreadsArgs *args = memnew(BakeThreadsArgs);
+	args->nav_mesh_instance = this;
+
+	bake_thread = Thread::create(_bake_navigation_mesh, args);
+	ERR_FAIL_COND(bake_thread == NULL);
+}
+
+void NavigationMeshInstance::_bake_done(Ref<NavigationMesh> p_nav_mesh) {
+	set_navigation_mesh(p_nav_mesh);
+	bake_thread = NULL;
+}
+
+String NavigationMeshInstance::get_configuration_warning() const {
+
+	if (!is_visible_in_tree() || !is_inside_tree())
+		return String();
+
+	if (!navmesh.is_valid()) {
+		return TTR("A NavigationMesh resource must be set or created for this node to work.");
+	}
+	const Spatial *c = this;
+	while (c) {
+
+		if (Object::cast_to<Navigation>(c))
+			return String();
+
+		c = Object::cast_to<Spatial>(c->get_parent());
+	}
+
+	return TTR("NavigationMeshInstance must be a child or grandchild to a Navigation node. It only provides navigation data.");
+}
+
+void NavigationMeshInstance::_bind_methods() {
+
+	ClassDB::bind_method(D_METHOD("set_navigation_mesh", "navmesh"), &NavigationMeshInstance::set_navigation_mesh);
+	ClassDB::bind_method(D_METHOD("get_navigation_mesh"), &NavigationMeshInstance::get_navigation_mesh);
+
+	ClassDB::bind_method(D_METHOD("set_enabled", "enabled"), &NavigationMeshInstance::set_enabled);
+	ClassDB::bind_method(D_METHOD("is_enabled"), &NavigationMeshInstance::is_enabled);
+
+	ClassDB::bind_method(D_METHOD("bake_navigation_mesh"), &NavigationMeshInstance::bake_navigation_mesh);
+	ClassDB::bind_method(D_METHOD("_bake_done", "nav_mesh"), &NavigationMeshInstance::_bake_done);
+
+	ADD_PROPERTY(PropertyInfo(Variant::OBJECT, "navmesh", PROPERTY_HINT_RESOURCE_TYPE, "NavigationMesh"), "set_navigation_mesh", "get_navigation_mesh");
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "enabled"), "set_enabled", "is_enabled");
+
+	ADD_SIGNAL(MethodInfo("navigation_mesh_changed"));
+	ADD_SIGNAL(MethodInfo("bake_done"));
```

Comment:
Should this signal be named `bake_finished`? I believe other signals throughout the engine have it in this way. I don't recall a signal with "done" for the name. (Only suggesting for constancy sake)

Reply:
There are other signals with the `_done` suffix, like:
- `_thumbnail_done`
- `_tree_thumbnail_done`
- `_file_list_thumbnail_done`
- `_preview_done`
- `_thread_done`
- `_request_done`

I didn't found any event with `_finish` suffix.
Despite this, do you think it would make more sense: `bake_finish`?

Reply:
@AndreaCatania We have signals that end with `completed` and `finished`, so we should use past tense, not present tense.

Reply:
@AndreaCatania The first one that pops in my mind is with animations. I believe audio is "finished" as well. I wasn't aware of the other signals you listed. Other than the `_request_done` one, but that one slipped my mind.

@Calinou If this is the case - does this call in the question some of the ones that they mentioned?

Reply:
My fault, in the fast check that I did this morning I searched for: `_finish"` instead of `_finished"`.
I've already changed it, it's `bake_finished` now.